### ISSUE TITLE
Dynamically Add A Page while creating a PDFDocument

### ISSUE THREAD

**Issue by harishchopra86** [techprimate/TPPDF#267]
I am using this library to generate a PDF of content added by the user. I am using: `let pdfDocument = PDFDocument(format: .a4)` and then adding text, table as required.

I wanted to know if there is a way to dynamically create a new page once content had reached the end of page one.
Please advise.

**Comment by philprime** [comment]
Hi, you can add a page break by calling `document.createNewPage()`.

This can be found [here](https://github.com/techprimate/TPPDF/blob/master/Documentation/Usage.md#page-break)

**Comment by harishchopra86** [comment]
Thanks for the prompt response @philprime 

However, I am having a different problem as follows:
1. I created a pdf document
2. I am adding text and spaces as per the content requirements.
3. As the content had passed the first page, it no more shows as second page is not generated.
4. As per your suggestion, we have `document.createNewPage()` method which will create a new page. I can add this method but I have no idea when my content would pass the first page or second page.

Basically, I have input fields for the user in which the user is entering content and I need to create pdf from that content. So the content is totally dynamic, in some cases, it would be small and in other cases, it would be quite large.

So wanted to know if there is a way to calculate when content had passed a page height or any other way to handle this situation.
Please advise.

**Comment by harishchopra86** [comment]
@philprime I am stuck on this for the last couple of days. Any advice on how should I handle this use case.

**Comment by philprime** [comment]
Can you provide an example to reproduce your current state/issue?

**Comment by harishchopra86** [comment]
@philprime Here is an example code which cuts out content after it had reached end of first page:
`
 func generateDocument() -> [PDFDocument] {
        let document = PDFDocument(format: .a4)

        let items = [
            "Simple text drawing",
            "Advanced text drawing using AttributedString",
            "Multi-layer rendering by simply setting the offset",
            "Fully calculated content sizing",
            "Automatic page wrapping",
            "Customizable pagination",
            "Fully editable header and footer",
            "Simple image positioning and rendering",
            "Image captions",
            "Simple text drawing",
            "Advanced text drawing using AttributedString",
            "Multi-layer rendering by simply setting the offset",
            "Fully calculated content sizing",
            "Automatic page wrapping",
            "Customizable pagination",
            "Fully editable header and footer",
            "Simple image positioning and rendering",
            "Image captions",
            "Simple text drawing",
            "Advanced text drawing using AttributedString",
            "Multi-layer rendering by simply setting the offset",
            "Fully calculated content sizing",
            "Automatic page wrapping",
            "Customizable pagination",
            "Fully editable header and footer",
            "Simple image positioning and rendering",
            "Image captions",
            "Simple text drawing",
            "Advanced text drawing using AttributedString",
            "Multi-layer rendering by simply setting the offset",
            "Fully calculated content sizing",
            "Automatic page wrapping",
            "Customizable pagination",
            "Fully editable header and footer",
            "Simple image positioning and rendering",
            "Image captions",
            "Simple text drawing",
            "Advanced text drawing using AttributedString",
            "Multi-layer rendering by simply setting the offset",
            "Fully calculated content sizing",
            "Automatic page wrapping",
            "Customizable pagination",
            "Fully editable header and footer",
            "Simple image positioning and rendering",
            "Image captions"
        ]

        // Simple bullet point list
        let featureList = PDFList(indentations: [
            (pre: 10.0, past: 20.0),
            (pre: 20.0, past: 20.0),
            (pre: 40.0, past: 20.0)
        ])

        // By adding the item first to a list item with the dot symbol, all of them will inherit it
        featureList
            .addItem(PDFListItem(symbol: .dot)
                .addItems(items.map({ item in
                    PDFListItem(content: item)
                })))
        document.add(list: featureList)

        document.add(space: 20)

        // Numbered list with unusual indentation
        let weirdIndentationList = PDFList(indentations: [
            (pre: 10.0, past: 20.0),
            (pre: 40.0, past: 30.0),
            (pre: 20.0, past: 50.0)
        ])

        weirdIndentationList
            .addItems(items.enumerated().map({ arg in
                PDFListItem(symbol: .numbered(value: "\(arg.offset + 1)"), content: arg.element)
            }))
        document.add(list: weirdIndentationList)

        return [document]
    }
`


In this example have added more content to the `ListExampleFactory` from example project.

**Comment by walery1409** [comment]
I have the same issue. Using PDFList, content is cut at the end of the page

**Comment by harishchopra86** [comment]
@philprime Any advice on this or any timeline for the feature?

**Comment by philprime** [comment]
Hi everyone, I just tested the sample to reproduce (thanks @harishchopra86).
Now I can see what the actual issue is:

<img width="627" alt="image" src="https://user-images.githubusercontent.com/30923007/113123835-340a7080-9215-11eb-9ab3-bb06bd7c50a6.png">

The issue is that the `PDFListObject` which is responsible for transforming the `PDFList` into a list of `PDFAttributedString` elements, doesn't add the page break as expected.

I'll try to see if I can find the issue

**Comment by walery1409** [comment]
> @philprime Any advice on this or any timeline for the feature?

I've solved this by adding NSAttributedStrings (with bullets) to PDFTable. It's not ideal solution but it's working.

**Comment by harishchopra86** [comment]
Thanks for the changes @philprime 
It is working for me when installing from `267-fixed-list-breaking` branch.

**Comment by philprime** [comment]
Great, I am creating a PR and merge it into `develop`

**Comment by philprime** [comment]
This should be fixed now in 2.3.5, please reopen if the issue still remains

**Comment by USBA** [comment]
@philprime, first of all, thanks so much for this super simple and easy-to-use library.

But I'm still facing the same issue for images.

Tested on:
- TPPDF version: 2.3.5
- Xcode 14.0 beta 3
- iOS 16 developer beta 3

Here's a simple, reproducible example.

```
// an array that contains 7 UIImages
var uiImgaes: [UIImage] = [
    UIImage(named: "1")!,
    UIImage(named: "2")!,
    UIImage(named: "3")!,
    UIImage(named: "4")!,
    UIImage(named: "5")!,
    UIImage(named: "6")!,
    UIImage(named: "7")!
]
    
func generatePDF() -> PDFDocument {
    let document = PDFDocument(format: .a4)

    // add each image in the array to the PDF
    for img in uiImgaes {
        let imageElement = PDFImage(image: img)
        document.add(image: imageElement)
    }
        
    return document
}
```

**Current behaviour:** 
Here's the generated PDF:
![IMG_9624](https://user-images.githubusercontent.com/19838220/180617618-61f065c9-4c9c-402f-a130-13ffc76d9b13.PNG)
The generated PDF only contains a single page and that page can only fit two images.

**Expected behaviour:** 
Dynamically create new pages when there're no space left for more images in the current page
